pubsub/google/cloud/pubsub_v1/subscriber/policy/base.py

```diff
@@ -0,0 +1,392 @@
+        computes how close the subscriber is to each of them, and returns
+        whichever value is higher. (It does not matter that we have lots of
+        running room on setting A if setting B is over.)
+
+        Returns:
+            float: The load value.
+        """
+        return max([
+            len(self.managed_ack_ids) / self.flow_control.max_messages,
+            self._bytes / self.flow_control.max_bytes,
+        ])
+
+    def ack(self, ack_id, time_to_ack=None, byte_size=None):
+        """Acknowledge the message corresponding to the given ack_id.
+
+        Args:
+            ack_id (str): The ack ID.
+            time_to_ack (int): The time it took to ack the message, measured
+                from when it was received from the subscription. This is used
+                to improve the automatic ack timing.
+            byte_size (int): The size of the PubSub message, in bytes.
+        """
+        # If we got timing information, add it to the histogram.
+        if time_to_ack is not None:
+            self.histogram.add(int(time_to_ack))
+
+        # Send the request to ack the message.
+        # However, if the consumer is inactive, then queue the ack_id here
+        # instead; it will be acked as part of the initial request when the
+        # consumer is started again.
+        if self._consumer.active:
+            request = types.StreamingPullRequest(ack_ids=[ack_id])
+            self._consumer.send_request(request)
+        else:
+            self._ack_on_resume.add(ack_id)
+
+        # Remove the message from lease management.
+        self.drop(ack_id=ack_id, byte_size=byte_size)
+
+    def call_rpc(self, request_generator):
+        """Invoke the Pub/Sub streaming pull RPC.
+
+        Args:
+            request_generator (Generator): A generator that yields requests,
+                and blocks if there are no outstanding requests (until such
+                time as there are).
+        """
+        return self._client.api.streaming_pull(request_generator)
+
+    def drop(self, ack_id, byte_size):
+        """Remove the given ack ID from lease management.
+
+        Args:
+            ack_id (str): The ack ID.
+            byte_size (int): The size of the PubSub message, in bytes.
+        """
+        # Remove the ack ID from lease management, and decrement the
+        # byte counter.
+        if ack_id in self.managed_ack_ids:
+            self.managed_ack_ids.remove(ack_id)
+            self._bytes -= byte_size
+            self._bytes = min([self._bytes, 0])
+
+        # If we have been paused by flow control, check and see if we are
+        # back within our limits.
+        #
+        # In order to not thrash too much, require us to have passed below
+        # the resume threshold (80% by default) of each flow control setting
+        # before restarting.
+        if self._paused and self._load < self.flow_control.resume_threshold:
```

Comment:
I feel like this snippet should be move to a separate function? `resume_from_load`?